# Практикум 3

1. ***Округление вверх и вниз***

Напишите программу, которая:
принимает число с плавающей запятой
выводит результат округления вверх и вниз

Пример вывода:

Введите число: 4.3

Округление вверх: 5

Округление вниз: 4

In [2]:
import math

num = float(input())
print(math.ceil(num))
print(math.floor(num))

 3.5


4
3


In [3]:
num = float(input())
print(int(num)+1)
print(int(num))

 3.5


4
3


2. Найти сумму числового ряда от 1 до введенного числа n

In [4]:
n = int(input())
i = 1
s = 0 
while i <= n:
    s += i
    i += 1
print(s)

 100


5050


Найти сумму четных чисел ряда


In [5]:
n = int(input())
i = 0
s = 0 
while i <= n:
    s += i
    i += 2
print(s)

 100


2550


In [6]:
n = int(input())
i = 0
s = 0 
while i <= n:
    if i%2 == 0 :
        s += i
    i += 1
print(s)

 100


2550


Найти сумму чисел числового ряда, которые делятся на 3 и 11

In [8]:
n = int(input())
i = 0
s = 0 
while i <= n:
    if i%3 == 0 and i%11 == 0:
        s += i
    i += 1
print("Итог:",s)

 100


Итог: 198


3. ***Счётчик факториала***
 
Напишите программу, которая рассчитывает факториал заданного числа 
(произведение всех целых чисел от 1 до самого числа) двумя способами.
Выведите оба результата и сравните равны ли они.

Пример вывода:

Введите число: 5

Факториал числа 5 первым способом: 120

Факториал числа 5 вторым способом: 120

Результаты равны? True

In [12]:
import math

n = int(input())
f = math.factorial(n)
print(str(n)+'!=',f)
#----------------------------
i = 1
p = 1 
while i <= n:
    p *= i
    i += 1

print(str(n)+'!=',p)

print(p == f)

 6


6!= 720
6!= 720
True


Для заданного целого числа N подсчитать количество четных чисел, меньше или равных N.


In [14]:
n = int(input())
i = 1
cnt = 0 
while i <= n:
    if i%2 == 0:
        cnt += 1
    i += 1
print(cnt)

 5


2


По данному натуральному числу N найдите наибольшую целую степень двойки, не превосходящую N. Выведите показатель степени и саму степень.


In [16]:
n = int(input())
p = 0
while 2**p <= n:
    p += 1
print('2',p-1)

 10


2 3


In [17]:
n = int(input())
p = 0
while True:
    if 2**p > n:
        break
    p += 1
print(p-1)

 10


3


In [18]:
n = int(input())
p = 0
while n//2 > 0:
    p += 1
    n //= 2
print(p)

 10


3


Дано целое число, не меньшее 2. Выведите его наименьший натуральный делитель, отличный от 1.


In [21]:
n = int(input()) #15

i = 2 # 2.....n
while True:
    if n % i == 0:
        print('НОД',i)
        break
    i += 1

 15


НОД 3


Определите сумму всех элементов последовательности, завершающейся числом 0. Числа, следующие за первым нулем, учитывать не нужно. Числа считываем с клавиатуры с помощью input()


In [23]:
s = 0
while True:
    num = int(input())
    if num == 0:
        break
    s += num
print(s)

 1
 2
 3
 4
 5
 0


15


In [24]:
s = 0
num = int(input())
while num != 0:
    s += num
    num = int(input())
print(s)

 1
 2
 3
 4
 5
 0


15


4. ***Симулятор кассы***

Напишите программу, которая принимает от пользователя цены товаров. Пользователь вводит цены по одному (ввод завершается при вводе нуля). 

Программа должна:

* Рассчитать и вывести общую сумму покупки, округленную до целого числа.
* Определить, какую сумму покупатель должен заплатить, если у него есть только купюры номиналом 100 евро.
* Рассчитать и вывести сдачу, которую нужно дать покупателю.

In [28]:
total_sum = 0
while True:
    price = float(input('Введите цену товара: (Для завершения ввода нажмите 0)'))
    if price == 0:
        print('Округленная сумма:',round(total_sum))
        break
    total_sum += price
cnt100 = total_sum//100 if total_sum%100 == 0  else  total_sum//100 + 1
print('Дай', int(cnt100),'по 100 евро')
print('Сдача:', cnt100*100 - total_sum)


Введите цену товара: (Для завершения ввода нажмите 0) 123
Введите цену товара: (Для завершения ввода нажмите 0) 39
Введите цену товара: (Для завершения ввода нажмите 0) 890
Введите цену товара: (Для завершения ввода нажмите 0) 0


Округленная сумма: 1052
Дай 11 по 100 евро
Сдача: 48.0


5. ***Учет доходов и расходов***

Напишите программу, которая принимает на вход положительные и отрицательные значения и останавливается получив ноль. Программа должна рассчитать сумму доходов (положительные) и расходов (отрицательные), а также итоговый баланс.

Пример вывода:

Введите доход, расход или '0' для получения итога: 15000

Введите доход, расход или '0' для получения итога: -5000

Введите доход, расход или '0' для получения итога: -3000

Введите доход, расход или '0' для получения итога: 0

Доходы: 15000

Расходы: 8000

Итоговый баланс: 7000


In [29]:
income = 0
cost = 0
while True:
    value = int(input('Введите число: (Для выхода используем 0)'))
    if value == 0:
        print('Доходы:',income)
        print('Доходы:',cost)
        break
    if value > 0:
        income += value
    if value < 0:
        cost += value
balans = income + cost
print('Итоговый баланс: ', balans)       

Введите число: (Для выхода используем 0) 1
Введите число: (Для выхода используем 0) 2
Введите число: (Для выхода используем 0) -3
Введите число: (Для выхода используем 0) 0


Доходы: 3
Доходы: -3
Итоговый баланс:  6


Определите среднее значение всех элементов последовательности, завершающейся числом 0. Среднее значение - сумма всех элементов, поделенная на их количество. 


Последовательность состоит из натуральных чисел и завершается числом 0. Определите значение наибольшего элемента последовательности.


Последовательность состоит из натуральных чисел и завершается числом 0. Определите индекс наибольшего элемента последовательности. Нумерация элементов начинается с нуля.


6. Вычислите n-е число ряда Фибоначчи с помощью цикла while. Что такое числа Фибоначчи?


7. ***Число Армстронга***

Напишите программу, которая запрашивает у пользователя целое число и проверяет, является ли оно числом Армстронга. Число Армстронга — это число, которое равно сумме своих цифр, которые были возведены в степень, равную количеству цифр в числе.
Например 153 = 1^3 + 5^3 + 3^3 = 1 + 125 + 27

Пример вывода:

Введите число: 153

Число 153 является числом Армстронга.
    
Введите число: 123

Число 123 не является числом Армстронга.


8. ***Угадай за 5 попыток***
    
Напишите программу, которая генерирует случайное число от 1 до 10 и просит пользователя угадать его за 5 попыток. Если пользователь не угадывает за 5 попыток, программа выводит загаданное число.

Пример вывода:

Попробуйте угадать число от 1 до 10: 10

Неверно! У вас осталось 4 попытки.
    
Попробуйте угадать число от 1 до 50: 3

Неверно! У вас осталось 3 попытки.
    
Попробуйте угадать число от 1 до 50: 7

Поздравляем! Вы угадали число.


9. ***Определение простого числа***
    
Напишите программу, которая запрашивать число у пользователя и определяет, является ли заданное число простым. 

Число является простым, если оно:
Больше 1.
Имеет только два делителя: единицу и само себя.

Пример вывода:
Введите число: 29
Число 29 является простым.
Введите число: 15
Число 15 не является простым.


10. ***Упорядоченность цифр***

Напишите программу, которая получает натуральное число и определяет, являются ли его цифры упорядоченными по возрастанию.

Пример вывода:
Введите число: 1347
Цифры упорядочены по возрастанию.
Введите число: 1523
Цифры не упорядочены по возрастанию.
